crates/red_knot_python_semantic/src/semantic_index/use_def.rs

```diff
@@ -323,37 +313,32 @@ impl<'db> UseDefMap<'db> {
         }
     }
 
-    pub(crate) fn public_declarations(
-        &self,
+    pub(crate) fn public_declarations<'map>(
+        &'map self,
         symbol: ScopedSymbolId,
-    ) -> DeclarationsIterator<'_, 'db> {
+    ) -> DeclarationsIterator<'map, 'db> {
         let declarations = self.public_symbols[symbol].declarations();
         self.declarations_iterator(declarations)
     }
 
-    pub(crate) fn has_public_declarations(&self, symbol: ScopedSymbolId) -> bool {
-        !self.public_symbols[symbol].declarations().is_empty()
-    }
-
-    fn bindings_iterator<'a>(
-        &'a self,
-        bindings: &'a SymbolBindings,
-    ) -> BindingWithConstraintsIterator<'a, 'db> {
+    fn bindings_iterator<'map>(
+        &'map self,
+        bindings: &'map SymbolBindings,
+    ) -> BindingWithConstraintsIterator<'map, 'db> {
         BindingWithConstraintsIterator {
             all_definitions: &self.all_definitions,
             all_constraints: &self.all_constraints,
-            inner: bindings.iter(),
+            inner: bindings.iter(&self.all_constraints, &self.visibility_constraints),
```

Comment:
> I see they are used in type inference, but they will be the same for every `BindingWithConstraints`; couldn't they could be accessed directly off the `BindingWithConstraintsIterator` as attributes (or directly off the use-def map?), rather than making every `BindingWithConstraints` two words larger?

Yes — thanks! I was about to clean this all up (it was in this state from a previous iteration), but I'm not sure if I would have spotted this. Changed it now.

> Or alternatively, not sure if `BindingWithConstraintsIterator::next` could itself encapsulate more of the logic that is currently in `bindings_ty`, and not require direct access to these IndexVecs in type inference. (Hmm, maybe this last is not a good idea, if it means that iterating over `BindingWithConstraintsIterator` would then require a db and doing type inference of visibility constraints.)

I think this might be similar to what I had in my first iteration. I basically had `VisibilityConstraintRef`erences and they would be transformed into `VisibilityConstraint`s in the `next` method. The `VisibilityConstraint`s then didn't contain any references to the `IndexMap` anymore, but this required the tree nodes to be boxed, which was far too slow.

Let me know if you had something else in mind.

Reply:
I think my third paragraph was a bad suggestion and I should have just left it out; it sounds like you already explored that design space. Happy if we can just avoid the extra references in every `BindingWithConstraints`.